File: 680-validPalindromeII/solution.py
class Solution(object):
    def validPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """

        
        start, end = 0, len(s)-1

        while start<end:
            if s[start]==s[end]:
                start+=1
                end-=1
            else:
                # 22%
                return self.isPalindrome(s[:start] + s[(start+1):]) or self.isPalindrome(s[:end] + s[(end+1):]) 
                # 56%
                return self.isPalindrome(s[start:end-1+1]) or self.isPalindrome(s[start+1:end+1]) 
        return True


    def isPalindrome(self, s):
        start, end = 0, len(s)-1

        while start<end:
            if s[start]==s[end]:
                start+=1
                end-=1
            else:
                return False
        return True

File: 680-validPalindromeII/solution.java
class Solution {
    public boolean validPalindrome(String s) {
        return isPalindrome(s, 0);    
    }
    
    // 60%
    public boolean isPalindrome(String s, int falseCount) {
        int start=0, end=s.length()-1;
        
        if (falseCount>1)   return false;

        while(start<end){
            if (s.charAt(start)==s.charAt(end)){
                start++;
                end--;

            } else {
                // System.out.println(s.substring(start, end)+" : "+s.substring(start+1, end+1));
                return isPalindrome(s.substring(start, end), falseCount+1) || isPalindrome(s.substring(start+1, end+1), falseCount+1);
            }
        }

        return true;

    }

}